numpy/random/_generator.pyx

```diff
@@ -3749,7 +3749,17 @@ cdef class Generator:
         pix = <double*>np.PyArray_DATA(parr)
         check_array_constraint(parr, 'pvals', CONS_BOUNDED_0_1)
         if kahan_sum(pix, d-1) > (1.0 + 1e-12):
-            raise ValueError("sum(pvals[:-1]) > 1.0")
+            msg = "sum(pvals[:-1]) > 1.0"
+            # When floating, but not float dtype, and close, improve the error
+            # 1.0001 works for float16 and float32
+            if (isinstance(pvals, np.ndarray) and
+                    pvals.dtype != float and
+                    np.issubdtype(pvals.dtype, np.floating) and
+                    pvals.sum() < 1.0001):
+                msg += (". pvals has been cast to double before checking "
+                        "the sum. Changes in precision when casting may "
+                        "produce violations even if pvals.sum() <= 1.")
```

Comment:
What are your thoughts on including `astype` in the message?
```suggestion
            msg = "sum(pvals[:-1]) > 1.0"
            # When floating, but not float dtype, and close, improve the error
            # 1.0001 works for float16 and float32
            if (isinstance(pvals, np.ndarray) and
                    pvals.dtype != float and
                    np.issubdtype(pvals.dtype, np.floating) and
                    pvals.sum() < 1.0001):
                msg += ("pvals[:-1].astype(np.float64).sum() > 1.0. @
                        "Changes in precision when casting may "
                        "produce violations even if pvals.sum() <= 1.")
            else:
                msg = "sum(pvals[:-1]) > 1.0"
```

Reply:
I like its precision. I kept a bit more test than you.